File: MPP/proiect/java/standalone/src/main/java/controller/Controller.java
package controller;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import model.Purchase;
import model.ShowData;
import services.IClientService;
import services.IServerService;

import java.util.Collection;

/**
 * Created by vitiv on 3/21/17.
 */
public class Controller implements IClientService {
    private IServerService serverService;
    private ObservableList<ShowData> dataModel;

    public ObservableList<ShowData> getDataModel() {
        return dataModel;
    }


    public Controller(IServerService serverService) {
        this.serverService = serverService;
        dataModel = FXCollections.observableArrayList();
        populateList();
    }

    public void addPurchase(int showId, String name, int quantity) {
        serverService.addPurchase(new Purchase(0, showId, name, quantity));
        populateList();
    }

    private void populateList() {
        dataModel.clear();
        Iterable<ShowData> tasks = serverService.getAll();
        tasks.forEach(x -> dataModel.add(x));
    }


    public boolean login(String username, String password) {
        return serverService.login(username, password, this);
    }

    @Override
    public void changesOccurred(Collection<ShowData> changedValues) {
        populateList();
    }
}

File: MPP/proiect/c#/Festival/Controllers/Controller/Controller.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Festival.Model;
using Festival.Service;

namespace Festival.Controller
{
    
    public class Controller: IClientService
    {
        private IServerService serv;
        public IBindingList ShowBindingList { get; }

        public Controller(IServerService serv)
        {
            this.serv = serv;
            ShowBindingList = new BindingList<ShowData>();           
        }

        private SynchronizationContext threadingContext;

        public void setThreadingContext(SynchronizationContext threadingContext)
        {
            // used for updating when ui is in another thread
            this.threadingContext = threadingContext;
        }

        public void AddPurchase(int showId, string clientName, int quantity)
        {
            serv.AddPurchase(showId, clientName, quantity);
        }

        public bool Login(string username, string password)
        {
            if (serv.Login(username, password, this))
            {
                ShowBindingList.Clear();
                foreach (var showData in serv.GetAll())
                {
                    ShowBindingList.Add(showData);
                }
                return true;
            }
            return false;
        }

        public void handleUpdate(IEnumerable<ShowData> changed)
        {
            // see on what pos is which id
            var asoc = new Dictionary<int, int>();
            for (var i = 0; i < ShowBindingList.Count; ++i)
                asoc[((ShowData) ShowBindingList[i]).Id] = i;
            // update corersponding items
            foreach (var show in changed)
                ShowBindingList[asoc[show.Id]] = show;
        }

        public void changesOccured(IEnumerable<ShowData> changed)
        {
            if (threadingContext != null)
            {
                threadingContext.Post(state =>
                {
                    this.handleUpdate((IEnumerable<ShowData>) state);
                }, changed);
            }
            else
            {
                this.handleUpdate(changed);
            } 
               
        }
    }
}
